In [2]:
# Import libraries

import pandas as pd 
import numpy as np 
import os 

In [3]:
# Create path 
path =  r'/Users/julie.spigner/Desktop/Kiva Data Analysis'

In [4]:
path

'/Users/julie.spigner/Desktop/Kiva Data Analysis'

In [5]:
# Import dataframes 
df_loans = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'kiva_loans.csv'), index_col = False)

In [6]:
df_loc = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'kiva_mpi_region_locations.csv'), index_col = False)

## df_loans

In [7]:
# View dataframe  
df_loans.head()

id  funded_amount  loan_amount             activity          sector  \
0  653051          300.0        300.0  Fruits & Vegetables            Food   
1  653053          575.0        575.0             Rickshaw  Transportation   
2  653068          150.0        150.0       Transportation  Transportation   
3  653063          200.0        200.0           Embroidery            Arts   
4  653084          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  partner_id                posted_time  \
0        Lahore      PKR       247.0  2014-01-01 06:12:39+00:00   
1        Lahore      PKR       247.0  2014-01-01 06:51:08+00:00   
2     Maynaguri      INR       334.0  2014-01-01 09:58:07+00:00   
3        Lahore      PKR       247.0  2014-01-01 08:03:11+00:00   
4  Abdul Hakeem      PKR       245.0  2014-01-01 11:53:19+00:00   

              disbursed_time                funded_time  term_in_months  \
0  2013-12-17 08:00:00+00:00  2014-01-02 10:06:32+00:00            12.0   
1  2013-12-17 08:00:00+00:00  2014-01-02 09:17:23+00:00            11.0   
2  2013-12-17 08:00:00+00:00  2014-01-01 16:01:36+00:00            43.0   
3  2013-12-24 08:00:00+00:00  2014-01-01 13:00:00+00:00            11.0   
4  2013-12-17 08:00:00+00:00  2014-01-01 19:18:51+00:00            14.0   

   lender_count                          tags borrower_genders  \
0            12                           NaN           female   
1            14                           NaN   female, female   
2             6  user_favorite, user_favorite           female   
3             8                           NaN           female   
4            16                           NaN           female   

  repayment_interval        date  
0          irregular  2014-01-01  
1          irregular  2014-01-01  
2             bullet  2014-01-01  
3          irregular  2014-01-01  
4            monthly  2014-01-01

In [9]:
# View descriptive stats  
df_loans.describe()

id  funded_amount    loan_amount     partner_id  \
count  6.712050e+05  671205.000000  671205.000000  657698.000000   
mean   9.932486e+05     785.995061     842.397107     178.199616   
std    1.966113e+05    1130.398941    1198.660073      94.247581   
min    6.530470e+05       0.000000      25.000000       9.000000   
25%    8.230720e+05     250.000000     275.000000     126.000000   
50%    9.927800e+05     450.000000     500.000000     145.000000   
75%    1.163653e+06     900.000000    1000.000000     204.000000   
max    1.340339e+06  100000.000000  100000.000000     536.000000   

       term_in_months   lender_count  
count   671205.000000  671205.000000  
mean        13.739022      20.590922  
std          8.598919      28.459551  
min          1.000000       0.000000  
25%          8.000000       7.000000  
50%         13.000000      13.000000  
75%         14.000000      24.000000  
max        158.000000    2986.000000

In [10]:
# Get info 
df_loans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 671205 entries, 0 to 671204
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  671205 non-null  int64  
 1   funded_amount       671205 non-null  float64
 2   loan_amount         671205 non-null  float64
 3   activity            671205 non-null  object 
 4   sector              671205 non-null  object 
 5   use                 666973 non-null  object 
 6   country_code        671197 non-null  object 
 7   country             671205 non-null  object 
 8   region              614405 non-null  object 
 9   currency            671205 non-null  object 
 10  partner_id          657698 non-null  float64
 11  posted_time         671205 non-null  object 
 12  disbursed_time      668809 non-null  object 
 13  funded_time         622874 non-null  object 
 14  term_in_months      671205 non-null  float64
 15  lender_count        671205 non-nul

In [11]:
# Check for missing values 
df_loans.isnull().sum()

id                         0
funded_amount              0
loan_amount                0
activity                   0
sector                     0
use                     4232
country_code               8
country                    0
region                 56800
currency                   0
partner_id             13507
posted_time                0
disbursed_time          2396
funded_time            48331
term_in_months             0
lender_count               0
tags                  171416
borrower_genders        4221
repayment_interval         0
date                       0
dtype: int64

Notes: There are missing values in 7 columns (use, region, partner_id, disbursed_time, funded_time, tags, borrower_genders). All datatypes of missing values are strings. 

In [13]:
# Check size of dataframe to see ratio of missing values compared to all records 
df_loans.shape

(671205, 20)

In [16]:
# Create a subset of missing values in df 
df_loans_nan = df_loans[df_loans['region'].isnull() == True] 

In [18]:
df_loans_nan.head(50)

id  funded_amount  loan_amount                   activity  \
5    1080148          250.0        250.0                   Services   
49    653418          450.0        450.0              General Store   
54    653380          225.0        225.0                Food Market   
67   1080150          125.0        125.0                     Energy   
70    653244         2000.0       2000.0                     Retail   
99   1080153         5000.0       5000.0      Food Production/Sales   
102   653399         3975.0       3975.0                 Food Stall   
111   653376          225.0        225.0                     Retail   
112   653256         1925.0       2400.0         Electronics Repair   
114  1080151          125.0        125.0                     Energy   
135   653384          850.0        850.0                 Blacksmith   
140   653375         2975.0       2975.0      Food Production/Sales   
145   653427         1200.0       1200.0          Personal Expenses   
157   653408         3400.0       3400.0             Clothing Sales   
165   653403         2850.0       2850.0                    Poultry   
170   653177         4250.0       4250.0                   Catering   
186   653253         2625.0       3000.0              Grocery Store   
195  1080149          125.0        125.0              Grocery Store   
200   653236         2000.0       2000.0              Grocery Store   
202   653360          500.0        500.0              General Store   
205   653389          125.0        125.0                    Poultry   
215   653374          500.0        500.0  Personal Housing Expenses   
222   653232         2400.0       2400.0              Grocery Store   
231  1080154        10000.0      10000.0      Food Production/Sales   
232   653226         3000.0       3000.0              Mobile Phones   
233   653221         3000.0       3000.0              Grocery Store   
240   653422         1025.0       1025.0                     Retail   
255   653425          300.0        300.0        Fruits & Vegetables   
273   653412          375.0        375.0             Clothing Sales   
277   653361          500.0        500.0                 Metal Shop   
302   653421          375.0        375.0      Food Production/Sales   
309   653259         2750.0       3000.0              Grocery Store   
311  1080152          125.0        125.0                     Energy   
318   653433         1500.0       1500.0           Furniture Making   
349   653228         3000.0       3000.0              Mobile Phones   
355   653426          300.0        300.0             Clothing Sales   
362   653365         1025.0       1025.0                     Bakery   
365   653239         1800.0       1800.0              Grocery Store   
374   653423          700.0        700.0                     Retail   
381   653756          500.0        500.0  Personal Housing Expenses   
387   653786          100.0        100.0                Home Energy   
392   653782          100.0        100.0                Home Energy   
396   653462          400.0        400.0                Agriculture   
397   653750          350.0        350.0  Personal Housing Expenses   
401   653788          100.0        100.0                Home Energy   
405   653770          100.0        100.0                Home Energy   
407   653796          700.0        700.0              General Store   
412   653513         2350.0       2350.0               Beauty Salon   
414   653775          725.0        725.0                Agriculture   
416   653599          175.0        175.0                     Retail   

            sector                                                use  \
5         Services  purchase leather for my business using ksh 20000.   
49          Retail                                to stock his store.   
54            Food  to purchase various seasonal items to resell: ...   
67        Services                purchase solar lanterns for resale.   
70          Retail  to 

Notes: This tells me that missing values in region column is not country-specific. In other words, missing labels of regions occur across multiple countries. 

In [19]:
# Find duplicates 
df_loans_dups = df_loans[df_loans.duplicated()]

In [20]:
# Check for duplicates 
df_loans_dups

Empty DataFrame
Columns: [id, funded_amount, loan_amount, activity, sector, use, country_code, country, region, currency, partner_id, posted_time, disbursed_time, funded_time, term_in_months, lender_count, tags, borrower_genders, repayment_interval, date]
Index: []

Notes: There are no duplicates. 

In [21]:
# View dataframe with more rows 
df_loans.head(100)

id  funded_amount  loan_amount               activity  \
0    653051          300.0        300.0    Fruits & Vegetables   
1    653053          575.0        575.0               Rickshaw   
2    653068          150.0        150.0         Transportation   
3    653063          200.0        200.0             Embroidery   
4    653084          400.0        400.0             Milk Sales   
..      ...            ...          ...                    ...   
95   653381          325.0        325.0            Food Market   
96   653107          325.0        325.0          General Store   
97   653217         2200.0       2200.0               Pharmacy   
98   653345          600.0        600.0    Fruits & Vegetables   
99  1080153         5000.0       5000.0  Food Production/Sales   

            sector                                                use  \
0             Food            To buy seasonal, fresh fruits to sell.    
1   Transportation  to repair and maintain the auto rickshaw used ...   
2   Transportation  To repair their old cycle-van and buy another ...   
3             Arts  to purchase an embroidery machine and a variet...   
4             Food                           to purchase one buffalo.   
..             ...                                                ...   
95            Food  to purchase more red oil, imported oil, rice, ...   
96          Retail  to buy additional stocks for her sari-sari store.   
97          Health        to buy medicines, sand and sacks of cement.   
98            Food  to purchase onions, tomatoes, vegetables, and ...   
99            Food  hire more staff to keep up with the volume of ...   

   country_code        country            region currency  partner_id  \
0            PK       Pakistan            Lahore      PKR       247.0   
1            PK       Pakistan            Lahore      PKR       247.0   
2            IN          India         Maynaguri      INR       334.0   
3            PK       Pakistan            Lahore      PKR       247.0   
4            PK       Pakistan      Abdul Hakeem      PKR       245.0   
..          ...            ...               ...      ...         ...   
95           LR        Liberia          Caldwell      LRD       182.0   
96           PH    Philippines  Plaridel-Cebulin      PHP       126.0   
97           TZ       Tanzania     Dar es Salaam      TZS        87.0   
98           KE          Kenya            Likoni      KES       164.0   
99           US  United States               NaN      USD         NaN   

                  posted_time             disbursed_time  \
0   2014-01-01 06:12:39+00:00  2013-12-17 08:00:00+00:00   
1   2014-01-01 06:51:08+00:00  2013-12-17 08:00:00+00:00   
2   2014-01-01 09:58:07+00:00  2013-12-17 08:00:00+00:00   
3   2014-01-01 08:03:11+00:00  2013-12-24 08:00:00+00:00   
4   2014-01-01 11:53:19+00:00  2013-12-17 08:00:00+00:00   
..                        ...                        ...   
95  2014-01-02 16:09:14+00:00  2013-12-18 08:00:00+00:00   
96  2014-01-02 01:01:10+00:00  2013-12-03 08:00:00+00:00   
97  2014-01-02 08:19:38+00:00  2013-12-17 08:00:00+00:00   
98  2014-01-02 13:55:18+00:00  2013-12-14 08:00:00+00:00   
99  2014-01-02 20:13:47+00:00  2014-03-10 22:53:12+00:00   

                  funded_time  term_in_months  lender_count  \
0   2014-01-02 10:06:32+00:00            12.0            12   
1   2014-01-02 09:17:23+00:00            11.0            14   
2   2014-01-01 16:01:36+00:00            43.0             6   
3   2014-01-01 13:00:00+00:00            11.0             8   
4   2014-01-01 19:18:51+00:00            14.0            16   
..                        ...             ...           ...   
95  2014-01-04 22:05:18+00:00            11.0            10   
96  2014-01-02 14:24:58+00:00            11.0            10   
97  2014-01-03 23:17:08+00:00             8.0            77   
98  2014-01-08 22:25:13+00:00            14.0            22   
99  2014-03-11 05:53:06+00:00            30.0   

In [23]:
# View names of columns in the dataframe 
df_loans.columns

Index(['id', 'funded_amount', 'loan_amount', 'activity', 'sector', 'use',
       'country_code', 'country', 'region', 'currency', 'partner_id',
       'posted_time', 'disbursed_time', 'funded_time', 'term_in_months',
       'lender_count', 'tags', 'borrower_genders', 'repayment_interval',
       'date'],
      dtype='object')

Notes: df_loans looks good. All datatypes are clarified. No critical missing values need to be addressed. All columns are labeled correctly. There are no duplicates. 

## df_loc

In [26]:
# View dataframe 
df_loc.head(50)

LocationName  ISO      country            region  \
0     Badakhshan, Afghanistan  AFG  Afghanistan        Badakhshan   
1        Badghis, Afghanistan  AFG  Afghanistan           Badghis   
2        Baghlan, Afghanistan  AFG  Afghanistan           Baghlan   
3          Balkh, Afghanistan  AFG  Afghanistan             Balkh   
4         Bamyan, Afghanistan  AFG  Afghanistan            Bamyan   
5       Daykundi, Afghanistan  AFG  Afghanistan          Daykundi   
6          Farah, Afghanistan  AFG  Afghanistan             Farah   
7         Faryab, Afghanistan  AFG  Afghanistan            Faryab   
8         Ghazni, Afghanistan  AFG  Afghanistan            Ghazni   
9           Ghor, Afghanistan  AFG  Afghanistan              Ghor   
10       Helmand, Afghanistan  AFG  Afghanistan           Helmand   
11         Herat, Afghanistan  AFG  Afghanistan             Herat   
12       Jawzjan, Afghanistan  AFG  Afghanistan           Jawzjan   
13         Kabul, Afghanistan  AFG  Afghanistan             Kabul   
14      Kandahar, Afghanistan  AFG  Afghanistan          Kandahar   
15        Kapisa, Afghanistan  AFG  Afghanistan            Kapisa   
16         Khost, Afghanistan  AFG  Afghanistan             Khost   
17       Kunarha, Afghanistan  AFG  Afghanistan           Kunarha   
18        Kunduz, Afghanistan  AFG  Afghanistan            Kunduz   
19       Laghman, Afghanistan  AFG  Afghanistan           Laghman   
20         Logar, Afghanistan  AFG  Afghanistan             Logar   
21     Nangarhar, Afghanistan  AFG  Afghanistan         Nangarhar   
22        Nimroz, Afghanistan  AFG  Afghanistan            Nimroz   
23     Nooristan, Afghanistan  AFG  Afghanistan         Nooristan   
24       Paktika, Afghanistan  AFG  Afghanistan           Paktika   
25        Paktya, Afghanistan  AFG  Afghanistan            Paktya   
26      Panjsher, Afghanistan  AFG  Afghanistan          Panjsher   
27        Parwan, Afghanistan  AFG  Afghanistan            Parwan   
28      Samangan, Afghanistan  AFG  Afghanistan          Samangan   
29     Sar-E-Pul, Afghanistan  AFG  Afghanistan         Sar-E-Pul   
30        Takhar, Afghanistan  AFG  Afghanistan            Takhar   
31       Urozgan, Afghanistan  AFG  Afghanistan           Urozgan   
32        Wardak, Afghanistan  AFG  Afghanistan            Wardak   
33         Zabul, Afghanistan  AFG  Afghanistan             Zabul   
34                        NaN  ALB      Albania               NaN   
35                        NaN  ARM      Armenia               NaN   
36                        NaN  AZE   Azerbaijan               NaN   
37  Bujumbura Mairie, Burundi  BDI      Burundi  Bujumbura Mairie   
38              Nord, Burundi  BDI      Burundi              Nord   
39        Centre-Est, Burundi  BDI      Burundi        Centre-Est   
40             Ouest, Burundi  BDI      Burundi             Ouest   
41               Sud, Burundi  BDI      Burundi               Sud   
42             Alibori, Benin  BEN        Benin           Alibori   
43             Atacora, Benin  BEN        Benin           Atacora   
44          Atlantique, Benin  BEN        Benin        Atlantique   
45              Borgou, Benin  BEN        Benin            Borgou   
46            Collines, Benin  BEN        Benin          Collines   
47              Couffo, Benin  BEN        Benin            Couffo   
48               Donga, Benin  BEN        Benin             Donga   
49            Littoral, Benin  BEN        Benin          Littoral   

               world_region    MPI                              geo  \
0                South Asia  0.387  (36.7347725, 70.81199529999999)   
1                South Asia  0.466         (35.1671339, 63.7695384)   
2                South Asia  0.300         (35.8042947, 69.2877535)   
3                South Asia  0.301         (36.7550603, 66.8975372)   
4                South Asia  0.325         (34.8100067, 67.8212104)   
5                South Asia  0.313          (33.669495, 66.0463534)   


In [27]:
# Check size of the dataframe 
df_loc.shape

(2772, 9)

Notes: This dataframe--df_loc--is much smaller compared to df_loans. 

In [29]:
# Gather information 
df_loc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2772 entries, 0 to 2771
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   LocationName  984 non-null    object 
 1   ISO           1008 non-null   object 
 2   country       1008 non-null   object 
 3   region        984 non-null    object 
 4   world_region  1008 non-null   object 
 5   MPI           984 non-null    float64
 6   geo           2772 non-null   object 
 7   lat           892 non-null    float64
 8   lon           892 non-null    float64
dtypes: float64(3), object(6)
memory usage: 195.0+ KB


Notes: All datatypes are labeled correctly. 

In [30]:
# Check for missing values 
df_loc.isnull().sum()

LocationName    1788
ISO             1764
country         1764
region          1788
world_region    1764
MPI             1788
geo                0
lat             1880
lon             1880
dtype: int64

Notes: There are multiple missing values in all columns except for one (geo). There are only 2,772 rows in the entire dataframe. This means more than half of the dataframe is missing. For reference: 1,386 is half of 2,772. However, the geo column is the most important one and can be used to fill in missing values in some columns. I'm going to take a closer look at the geo column based on missing values in region column to see if this is true. 

In [31]:
# Create a subset to find missing values 
df_loc_nan = df_loc[df_loc['region'].isnull() == True]

In [33]:
df_loc_nan.head(25)

LocationName  ISO                                     country region  \
34            NaN  ALB                                     Albania    NaN   
35            NaN  ARM                                     Armenia    NaN   
36            NaN  AZE                                  Azerbaijan    NaN   
74            NaN  BIH                      Bosnia and Herzegovina    NaN   
118           NaN  BRB                                    Barbados    NaN   
234           NaN  DZA                                     Algeria    NaN   
413           NaN  IND                                       India    NaN   
449           NaN  KAZ                                  Kazakhstan    NaN   
458           NaN  KGZ                                  Kyrgyzstan    NaN   
510           NaN  LCA                                 Saint Lucia    NaN   
535           NaN  MDA                        Moldova, Republic of    NaN   
558           NaN  MDV                                    Maldives    NaN   
559           NaN  MEX                                      Mexico    NaN   
560           NaN  MKD  Macedonia, The former Yugoslav Republic of    NaN   
582           NaN  MNE                                  Montenegro    NaN   
781           NaN  PSE                        Palestine, State ofa    NaN   
837           NaN  SOM                                     Somalia    NaN   
838           NaN  SRB                                      Serbia    NaN   
907           NaN  THA                                    Thailand    NaN   
913           NaN  TKM                                Turkmenistan    NaN   
932           NaN  TUN                                     Tunisia    NaN   
952           NaN  UKR                                     Ukraine    NaN   
965           NaN  VUT                                     Vanuatu    NaN   
987           NaN  ZAF                                South Africa    NaN   
1008          NaN  NaN                                         NaN    NaN   

                     world_region  MPI               geo  lat  lon  
34        Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
35        Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
36        Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
74        Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
118   Latin America and Caribbean  NaN  (1000.0, 1000.0)  NaN  NaN  
234                   Arab States  NaN  (1000.0, 1000.0)  NaN  NaN  
413                    South Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
449       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
458       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
510   Latin America and Caribbean  NaN  (1000.0, 1000.0)  NaN  NaN  
535       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
558                    South Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
559   Latin America and Caribbean  NaN  (1000.0, 1000.0)  NaN  NaN  
560       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
582       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
781                   Arab States  NaN  (1000.0, 1000.0)  NaN  NaN  
837                   Arab States  NaN  (1000.0, 1000.0)  NaN  NaN  
838       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
907     East Asia and the Pacific  NaN  (1000.0, 1000.0)  NaN  NaN  
913       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
932                   Arab States  NaN  (1000.0, 1000.0)  NaN  NaN  
952       Europe and Central Asia  NaN  (1000.0, 1000.0)  NaN  NaN  
965     East Asia and the Pacific  NaN  (1000.0, 1000.0)  NaN  NaN  
987            Sub-Saharan Africa  NaN  (1000.0, 1000.0)  NaN  NaN  
1008                          NaN  NaN  (1000.0, 1000.0)  NaN  NaN

Notes: Apparently (1000.0, 1000.0) is considered a missing value in geo column. Check to see how many values in geo column with records of this. 

In [34]:
# Find specific cells in a column 
df_loc.loc[df_loc['geo'] == '(1000.0, 1000.0)']

LocationName  ISO       country region             world_region  \
34                   NaN  ALB       Albania    NaN  Europe and Central Asia   
35                   NaN  ARM       Armenia    NaN  Europe and Central Asia   
36                   NaN  AZE    Azerbaijan    NaN  Europe and Central Asia   
51          Ouðmð, Benin  BEN         Benin  Ouðmð       Sub-Saharan Africa   
63    Nord, Burkina Faso  BFA  Burkina Faso   Nord       Sub-Saharan Africa   
...                  ...  ...           ...    ...                      ...   
2767                 NaN  NaN           NaN    NaN                      NaN   
2768                 NaN  NaN           NaN    NaN                      NaN   
2769                 NaN  NaN           NaN    NaN                      NaN   
2770                 NaN  NaN           NaN    NaN                      NaN   
2771                 NaN  NaN           NaN    NaN                      NaN   

        MPI               geo  lat  lon  
34      NaN  (1000.0, 1000.0)  NaN  NaN  
35      NaN  (1000.0, 1000.0)  NaN  NaN  
36      NaN  (1000.0, 1000.0)  NaN  NaN  
51    0.238  (1000.0, 1000.0)  NaN  NaN  
63    0.581  (1000.0, 1000.0)  NaN  NaN  
...     ...               ...  ...  ...  
2767    NaN  (1000.0, 1000.0)  NaN  NaN  
2768    NaN  (1000.0, 1000.0)  NaN  NaN  
2769    NaN  (1000.0, 1000.0)  NaN  NaN  
2770    NaN  (1000.0, 1000.0)  NaN  NaN  
2771    NaN  (1000.0, 1000.0)  NaN  NaN  

[1880 rows x 9 columns]

In [35]:
# Command pandas not to assign any options regarding the max number of rows to display 
pd.options.display.max_rows = None 

In [36]:
# Find specific cells in a column 
df_loc.loc[df_loc['geo'] == '(1000.0, 1000.0)']

LocationName  ISO  \
34                                                  NaN  ALB   
35                                                  NaN  ARM   
36                                                  NaN  AZE   
51                                         Ouðmð, Benin  BEN   
63                                   Nord, Burkina Faso  BFA   
74                                                  NaN  BIH   
86              Potosð, Bolivia, Plurinational State of  BOL   
93                                        Amapð, Brazil  BRA   
98                               Espðrito Santo, Brazil  BRA   
99                                        Goiðs, Brazil  BRA   
104                                      Paranð, Brazil  BRA   
105                                     Paraðba, Brazil  BRA   
106                                        Parð, Brazil  BRA   
108                                       Piauð, Brazil  BRA   
118                                                 NaN  BRB   
141              Nana-Mambðrð, Central African Republic  CAF   
143               Ouham-Pendð, Central African Republic  CAF   
145                      Kðmo, Central African Republic  CAF   
146              Nana-Grðbizi, Central African Republic  CAF   
158                           Centre-Est, Cote d'Ivoire  CIV   
160                         Centre-Ouest, Cote d'Ivoire  CIV   
165                     Sud sans Abidjan, Cote d'Ivoire  CIV   
169                     Centre (sans Yaoundð), Cameroon  CMR   
172                              Extrðme-Nord, Cameroon  CMR   
179                                   Yaoundð, Cameroon  CMR   
183         ðquateur, Congo, Democratic Republic of the  COD   
184   Kasað-Occidental, Congo, Democratic Republic o...  COD   
185   Kasað-Oriental, Congo, Democratic Republic of the  COD   
188        Nord-Kivu, Congo, Democratic Republic of the  COD   
193                        Lðkoumou, Congo, Republic of  COG   
194                         Bouenza, Congo, Republic of  COG   
211                Caldas, Risaralda, Quindio, Colombia  COL   
212                    Tolima, Huila, Caqueta, Colombia  COL   
214                 Valle Sin Cali Ni Litoral, Colombia  COL   
215                Cauca Y Nariðo Sin Litoral, Colombia  COL   
220             Ndzuwani (Ndzouani or Anjouan), Comoros  COM   
223                           Other Districts, Djibouti  DJI   
224                  Cibao Nordeste, Dominican Republic  DOM   
234                                                 NaN  DZA   
245                                      Sharkia, Egypt  EGY   
277                                  Haut-Ogoouð, Gabon  GAB   
278                                 Moyen-Ogoouð, Gabon  GAB   
279                                      Ngounið, Gabon  GAB   
281           Ogoouð Maritime (sans Port-Gentil), Gabon  GAB   
282                                Ogoouð-Ivindo, Gabon  GAB   
283                                  Ogoouð-Lolo, Gabon  GAB   
284                                  Woleu-Nðtem, Gabon  GAB   
295                                        Bokð, Guinea  GIN   
300                                        Labð, Guinea  GIN   
302                                  N'Zðrðkorð, Guinea  GIN   
307                                     Kerewan, Gambia  GMB   
319                                  Sab, Guinea-Bissau  GNB   
345                            Demerara-Mahaica, Guyana  GUY   
352                                 Atlðntida, Honduras  HND   
353                                     Colðn, Honduras  HND   
356                                    Cortðs, Honduras  HND   
358                                El Paraðso, Honduras  HND   
359                         Francisco Morazðn, Honduras  HND   
361                                  Intibucð, Honduras  HND   
362                         Islas de la Bahða, Honduras  HND   
367                             Santa Bðrbara, Honduras  HND   
370              Aire Mðtropolitaine/Reste-Ouest, Haiti  HTI   
413        

Notes: So it looks like there are 1,880 missing values in geo column. The missing values vary across regions and are not location specific. Even though missing values are present in the geo column (and also in lat and long columns), there are enough information in other columns to use to provide location based information (ex: ISO, country, and world_region).

# df_loan_theme

In [38]:
# Import dataframe 
df_loan_theme = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'loan_theme_ids.csv'), index_col = False)

In [39]:
# Check size of dataframe 
df_loan_theme.shape

(779092, 4)

In [43]:
# View dataframe
df_loan_theme.head(30)

id    Loan Theme ID          Loan Theme Type  Partner ID
0   638631  a1050000000skGl                  General       151.0
1   640322  a1050000000skGl                  General       151.0
2   641006  a1050000002X1ij         Higher Education       160.0
3   641019  a1050000002X1ij         Higher Education       160.0
4   641594  a1050000002VbsW  Subsistence Agriculture       336.0
5   642256  a1050000000T3oX          Extreme Poverty       217.0
6   642311  a1050000000snTK                  General        57.0
7   642353  a1050000000T3oX          Extreme Poverty       217.0
8   642386  a1050000000wf0f                  General       133.0
9   642429  a1050000000wezJ                  General        48.0
10  642836  a1050000000wf0f                  General       133.0
11  643267  a1050000000T3oX          Extreme Poverty       217.0
12  643288  a1050000000wezk                  General        77.0
13  643402  a1050000000wezk                  General        77.0
14  643414  a1050000002VkWz              Underserved       334.0
15  643435  a1050000002VkWz              Underserved       334.0
16  643447  a1050000000shPp          Extreme Poverty        40.0
17  643557  a1050000000skGl                  General       151.0
18  643761  a1050000000wf1a                  General       205.0
19  644101  a1050000000snTK                  General        57.0
20  644518  a1050000000wezq                  General        85.0
21  645222  a1050000000SFvX      Mobile Transactions       219.0
22  645377  a1050000000wf0J                  General       112.0
23  645386  a1050000000wf0J                  General       112.0
24  645758  a1050000000wf19                  General       169.0
25  645781  a1050000000wezq                  General        85.0
26  645846  a1050000000wf19                  General       169.0
27  645918  a1050000000wf19                  General       169.0
28  645927  a1050000000wf19                  General       169.0
29  645953  a1050000000wf19                  General       169.0

In [44]:
# Check descriptive stats 
df_loan_theme.describe()

id     Partner ID
count  7.790920e+05  764279.000000
mean   1.047475e+06     180.825840
std    2.282538e+05      97.914029
min    6.386310e+05       9.000000
25%    8.499768e+05     126.000000
50%    1.046528e+06     145.000000
75%    1.244768e+06     204.000000
max    1.444243e+06     557.000000

In [45]:
# Get info 
df_loan_theme.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 779092 entries, 0 to 779091
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   id               779092 non-null  int64  
 1   Loan Theme ID    764279 non-null  object 
 2   Loan Theme Type  764279 non-null  object 
 3   Partner ID       764279 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 23.8+ MB


In [46]:
# Check for missing values 
df_loan_theme.isnull().sum()

id                     0
Loan Theme ID      14813
Loan Theme Type    14813
Partner ID         14813
dtype: int64

Notes: There are 14,813 missing values in 3 columns. There's a total of 779,092 records. This is relatively small. All of the missing values should be removed since there's no valuable information to fill in the missing values from one column to another except for id.

In [48]:
# Check frequency 
df_loan_theme['Loan Theme Type'].value_counts(dropna = False)

General                                                380693
Underserved                                             62575
Agriculture                                             40108
Rural Inclusion                                         21228
Water                                                   20131
Higher Education                                        17728
General (Women)                                         16566
Vulnerable Populations                                  16354
NaN                                                     14813
Green                                                   13913
Water and Sanitation                                    12032
Startup                                                  9887
Subsistence Agriculture and Solar Lanterns               8597
Youth                                                    8444
Subsistence Agriculture                                  8247
Vulnerable Women                                         8184
FUSAI   

Notes: Seems like some of the loan theme types are women-specific. Create a flag to note loan theme type to idenify women specific groups.

In [53]:
# Drop rows with missing values 
df_loan_theme.dropna(inplace = True)

In [54]:
# Check size of dataframe 
df_loan_theme.shape

(764279, 4)

Notes: Missing values have been successfully dropped from the dataframe. For reference: 779,092 - 14,813 = 764,279. 

## df_loan_theme_region

In [58]:
# Import dataframe 
df_loan_theme_region = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'loan_themes_by_region.csv'), index_col = False)

In [59]:
# View dataframe 
df_loan_theme_region.head(20)

Partner ID               Field Partner Name                       sector  \
0            9  KREDIT Microfinance Institution  General Financial Inclusion   
1            9  KREDIT Microfinance Institution  General Financial Inclusion   
2            9  KREDIT Microfinance Institution  General Financial Inclusion   
3            9  KREDIT Microfinance Institution  General Financial Inclusion   
4            9  KREDIT Microfinance Institution  General Financial Inclusion   
5            9  KREDIT Microfinance Institution  General Financial Inclusion   
6            9  KREDIT Microfinance Institution  General Financial Inclusion   
7            9  KREDIT Microfinance Institution  General Financial Inclusion   
8            9  KREDIT Microfinance Institution  General Financial Inclusion   
9            9  KREDIT Microfinance Institution  General Financial Inclusion   
10           9  KREDIT Microfinance Institution  General Financial Inclusion   
11           9  KREDIT Microfinance Institution  General Financial Inclusion   
12           9  KREDIT Microfinance Institution  General Financial Inclusion   
13           9  KREDIT Microfinance Institution  General Financial Inclusion   
14           9  KREDIT Microfinance Institution  General Financial Inclusion   
15           9  KREDIT Microfinance Institution  General Financial Inclusion   
16           9  KREDIT Microfinance Institution  General Financial Inclusion   
17           9  KREDIT Microfinance Institution  General Financial Inclusion   
18           9  KREDIT Microfinance Institution  General Financial Inclusion   
19           9  KREDIT Microfinance Institution  General Financial Inclusion   

      Loan Theme ID         Loan Theme Type   country forkiva  \
0   a1050000000slfi        Higher Education  Cambodia      No   
1   a10500000068jPe  Vulnerable Populations  Cambodia      No   
2   a1050000000slfi        Higher Education  Cambodia      No   
3   a10500000068jPe  Vulnerable Populations  Cambodia      No   
4   a1050000002X1Uu              Sanitation  Cambodia      No   
5   a1050000000slfi        Higher Education  Cambodia      No   
6   a1050000007VvXr      Solar Home Systems  Cambodia      No   
7   a1050000000weyk                 General  Cambodia      No   
8   a1050000007VvXr      Solar Home Systems  Cambodia      No   
9   a1050000000slfi        Higher Education  Cambodia      No   
10  a10500000068jPe  Vulnerable Populations  Cambodia      No   
11  a1050000002X1Uu              Sanitation  Cambodia      No   
12  a1050000002X1Uu              Sanitation  Cambodia      No   
13  a1050000000slfi        Higher Education  Cambodia      No   
14  a1050000000weyk                 General  Cambodia      No   
15  a1050000000slfi        Higher Education  Cambodia      No   
16  a1050000000weyk                 General  Cambodia      No   
17  a10500000068jPe  Vulnerable Populations  Cambodia      No   
18  a1050000000slfi        Higher Education  Cambodia      No   
19  a1050000000slfi        Higher Education  Cambodia      No   

                      region          geocode_old  ISO  ...  amount  \
0           Banteay Meanchey       (13.75, 103.0)  KHM  ...     450   
1        Battambang Province                  NaN  KHM  ...   20275   
2        Battambang Province                  NaN  KHM  ...    9150   
3      Kampong Cham Province        (12.0, 105.5)  KHM  ...  604950   
4      Kampong Cham Province        (12.0, 105.5)  KHM  ...     275   
5      Kampong Cham Province        (12.0, 105.5)  KHM  ...   62225   
6      Kampong Cham Province        (12.0, 105.5)  KHM  ...    1300   
7   Kampong Chhnang Province        (12.0, 104.5)  KHM  ...  237175   
8   Kampong Chhnang Province        (12.0, 104.5)  KHM  ...    3050   
9   Kampong Chhnang Province        (12.0, 104.5)  KHM  ...   31425   
10  Kampong Chhnang Province        (12.0, 104.5)  KHM  ...  323950   
11  Kampong Chhnang Province        (12.0, 104.5)  KHM  ...    1000   
12     Kampong Speu Province      

Notes: Seems like there are multiple columns. 

In [60]:
# Gather info 
df_loan_theme_region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15736 entries, 0 to 15735
Data columns (total 21 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Partner ID          15736 non-null  int64  
 1   Field Partner Name  15736 non-null  object 
 2   sector              15736 non-null  object 
 3   Loan Theme ID       15736 non-null  object 
 4   Loan Theme Type     15736 non-null  object 
 5   country             15736 non-null  object 
 6   forkiva             15736 non-null  object 
 7   region              15736 non-null  object 
 8   geocode_old         1200 non-null   object 
 9   ISO                 15722 non-null  object 
 10  number              15736 non-null  int64  
 11  amount              15736 non-null  int64  
 12  LocationName        15736 non-null  object 
 13  geocode             13662 non-null  object 
 14  names               13661 non-null  object 
 15  geo                 15736 non-null  object 
 16  lat 

In [61]:
# Check size of the dataframe 
df_loan_theme_region.shape

(15736, 21)

In [62]:
# Check for missing values 
df_loan_theme_region.isnull().sum()

Partner ID                0
Field Partner Name        0
sector                    0
Loan Theme ID             0
Loan Theme Type           0
country                   0
forkiva                   0
region                    0
geocode_old           14536
ISO                      14
number                    0
amount                    0
LocationName              0
geocode                2074
names                  2075
geo                       0
lat                    2074
lon                    2074
mpi_region               14
mpi_geo                6065
rural_pct              1392
dtype: int64

Note: There are multiple missing values in geocode_old column. Perhaps this column should be dropped. 

In [64]:
# Check for duplicates 
df_loan_theme_region_dups = df_loan_theme_region[df_loan_theme_region.duplicated()]

In [65]:
df_loan_theme_region_dups

Empty DataFrame
Columns: [Partner ID, Field Partner Name, sector, Loan Theme ID, Loan Theme Type, country, forkiva, region, geocode_old, ISO, number, amount, LocationName, geocode, names, geo, lat, lon, mpi_region, mpi_geo, rural_pct]
Index: []

[0 rows x 21 columns]

Notes: There are no duplicates in the dataframe. 

In [66]:
# Export dataframes
df_loans.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_loans_cleaned.pkl'))

In [67]:
df_loc.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_loc_cleaned.pkl'))

In [68]:
df_loan_theme.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_loan_theme_cleaned.pkl'))

In [70]:
df_loan_theme_region.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_loan_theme_region_cleaned.pkl'))